## Dependencies for RNA sequencing workshop

## Description:
To follow along with the practice in this course, you will use command line tools and R installed on your computer. Additionally, you will have to install the softwares and R libraries we will be using.

## Full list of Dependencies:
- Download [Raw data for this course](https://tufts.box.com/v/intro-to-RNA-seq-material)

- Command line tools for preprocessing reads and read alignment:
[FASTQC](https://www.bioinformatics.babraham.ac.uk/projects/fastqc/)
[STAR](https://github.com/alexdobin/STAR)
[samtools](http://www.htslib.org/download/)
[subread](http://subread.sourceforge.net)

 - R and its libraries for differential expression analysis:
[R studio](https://www.google.com/search?client=safari&rls=en&q=r+studio&ie=UTF-8&oe=UTF-8)

Bioconductr libraries (check out each library for installation guide):
1. [DESeq2](https://bioconductor.org/packages/release/bioc/html/DESeq2.html)
2. [vsn](https://bioconductor.org/packages/release/bioc/html/vsn.html)
3. [DEGreport](https://bioconductor.org/packages/release/bioc/html/DEGreport.html)
4. [org.Sc.sgd.db](https://bioconductor.org/packages/release/data/annotation/html/org.Sc.sgd.db.html)
5. [clusterProfiler](https://bioconductor.org/packages/release/bioc/html/clusterProfiler.html)

Regular libraries (install by entering `install.packages("name")`):
6. [tidyverse](https://ggplot2.tidyverse.org)
7. [ggplot2](https://ggplot2.tidyverse.org)
8. [dplyr](https://dplyr.tidyverse.org)
9. [ggrepel](https://cran.r-project.org/web/packages/ggrepel/vignettes/ggrepel.html#installation)
10. pheatmap
